nmtwizard/preprocess/consumer.py

```diff
@@ -16,57 +16,72 @@ def __call__(self, tu_batch):
         raise NotImplementedError()
 
 
+def _ingest_tokens(subword_learner, tu_side):
+    if not tu_side.tokens:
+        return
+    for part in tu_side.tokens:
+        for token in part:
+            # This method ignores annotations and placeholder tokens.
+            subword_learner.ingest_token(token)
+
+
 class SubwordLearner(Consumer):
     """SubwordLearner class stores, learns and writes subword models."""
 
     def __init__(self, config, result_dir, tok_step):
 
         self._result_dir = result_dir
 
+        self._subword_info = {}
         self._subword_learners = {}
 
         self._tok_step = tok_step
-        tok_config = config['preprocess'][tok_step]
-
-        opt_multi = tok_config.get('multi', {}).get('build_subword')
-        opt_source = tok_config.get('source', {}).get('build_subword')
-        opt_target = tok_config.get('target', {}).get('build_subword')
 
-        if opt_multi:
-            self._subword_learners['multi'] = tokenizer.make_subword_learner(opt_multi, result_dir)
-        if opt_source:
-            self._subword_learners['source'] = tokenizer.make_subword_learner(opt_source, result_dir)
-        if opt_target:
-            self._subword_learners['target'] = tokenizer.make_subword_learner(opt_target, result_dir)
+        for side in ('multi', 'source', 'target'):
+            tokenization_config = config['preprocess'][tok_step].get(side)
+            if tokenization_config is None:
+                continue
+            subword_config = tokenization_config.get('build_subword')
+            if subword_config is None:
+                continue
+            # The subword learner needs to be aware of the tokenizer annotations
+            # to properly ignore them.
+            learner_info = tokenizer.make_subword_learner(
+                subword_config,
+                result_dir,
+                tokenizer=tokenizer.build_tokenizer(tokenization_config))
+            self._subword_info[side] = learner_info
+            self._subword_learners[side] = learner_info['learner']
+
+        self._source_learner = self._subword_learners.get('source')
```

Comment:
Why do we need these extra references ?

Reply:
They save repetitive hash table lookups in the `__call__` method.

Reply:
OK, thanks for explanation!